jxl/src/render/mod.rs

```diff
@@ -3,103 +3,106 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
+use std::{any::Any, marker::PhantomData};
+
 use crate::{
     error::Result,
     image::{ImageDataType, ImageRectMut},
 };
 
+mod internal;
+mod simple_pipeline;
+
+use internal::RenderPipelineStageInfo;
+
+// pub use simple_pipeline::SimpleRenderPipeline;
+
 /// These are the only stages that are assumed to have observable effects, i.e. calls to process_row
 /// for other stages may be omitted if it can be shown they can't affect any Input stage process_row
 /// call that happens inside image boundaries.
-pub trait RenderPipelineInputStage<InputT: ImageDataType> {
-    fn uses_channel(&self, c: usize) -> bool;
-
-    /// `row` has as many entries as the number of used channels for the stage, i.e. the channels
-    /// for which `uses_channel` returns true.
-    fn process_row_chunk(&mut self, position: (usize, usize), xsize: usize, row: &[&[InputT]]);
+/// For these stages, `process_row_chunk` receives read-only access to each row in each channel.
+pub struct RenderPipelineInputStage<InputT: ImageDataType> {
+    _phantom: PhantomData<InputT>,
```

Comment:
What are these phantoms for?

Reply:
Rust needs every generic parameter to be used in the struct definition, `PhantomData` is a lang-item ZST that is usually used for that purpose